File: apps/agent/src/agents/workspaceAgent/tools/listInbox.ts
import { FunctionTool } from '@google/adk';
import { z } from 'zod';
import type { WorkspaceTokensStore } from '../../../storage/workspaceTokens.js';
import type { ExecFileLike } from '../gwsExec.js';
import { type RunGwsLogEvent, runGws } from '../runGws.js';

/**
 * `list_inbox` — sub-agent's read tool for "what's in the inbox right
 * now". Returns bare summaries (id, threadId, snippet) — no body.
 * The sub-agent calls `get_message` per id when it needs the full
 * content (typically only for non-trivial messages it wants to triage).
 */

export const LIST_INBOX_TOOL_NAME = 'list_inbox';

export interface CreateListInboxToolDeps {
  store: WorkspaceTokensStore;
  uid: string;
  execFile?: ExecFileLike;
  log?: (event: RunGwsLogEvent) => void;
}

export interface InboxMessageSummary {
  id: string;
  threadId: string;
  snippet: string;
}

export type ListInboxResult =
  | { status: 'ok'; messages: InboxMessageSummary[]; truncated?: boolean }
  | { status: 'error'; code: string; message: string };

const parameters = z.object({
  unreadOnly: z.boolean().optional().describe('When true, restrict to unread messages only.'),
  since: z
    .string()
    .optional()
    .describe(
      'Gmail-style relative window (e.g. "1d", "12h", "3d"). Default "1d" — last 24 hours.',
    ),
  limit: z
    .number()
    .int()
    .positive()
    .max(50)
    .optional()
    .describe('Maximum number of messages to return (1–50). Default 15.'),
});

interface ListResponse {
  messages?: { id?: string; threadId?: string }[];
}

interface MessageMetadata {
  id?: string;
  threadId?: string;
  snippet?: string;
}

export function createListInboxTool(deps: CreateListInboxToolDeps): FunctionTool {
  const { store, uid, execFile, log } = deps;

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: LIST_INBOX_TOOL_NAME,
    description:
      'List recent inbox messages as id+threadId+snippet summaries (no body). Use get_message per id to read full content. Read-only.',
    parameters,
    execute: async (input: unknown): Promise<ListInboxResult> => {
      const args = input as { unreadOnly?: boolean; since?: string; limit?: number };
      const since = args.since ?? '1d';
      const maxResults = args.limit ?? 15;
      const q = `${args.unreadOnly ? 'is:unread ' : ''}label:INBOX newer_than:${since}`.trim();

      const listResult = await runGws({
        store,
        uid,
        toolName: LIST_INBOX_TOOL_NAME,
        service: 'gmail',
        resource: 'users.messages',
        method: 'list',
        params: { userId: 'me', q, maxResults },
        execFile,
        log,
      });
      if (listResult.status === 'error') {
        return { status: 'error', code: listResult.code, message: listResult.message };
      }
      const listBody = (listResult.body as ListResponse | null) ?? {};
      const ids = (listBody.messages ?? [])
        .map((m) => m.id)
        .filter((id): id is string => typeof id === 'string' && id.length > 0);
      if (ids.length === 0) {
        return { status: 'ok', messages: [] };
      }

      // Fetch metadata-format snippets in parallel for the listed ids.
      // metadata format avoids base64 bodies entirely — we only need snippets here.
      const detailResults = await Promise.all(
        ids.map((id) =>
          runGws({
            store,
            uid,
            toolName: LIST_INBOX_TOOL_NAME,
            service: 'gmail',
            resource: 'users.messages',
            method: 'get',
            params: { userId: 'me', id, format: 'metadata' },
            execFile,
            log,
          }),
        ),
      );

      const messages: InboxMessageSummary[] = [];
      for (const detail of detailResults) {
        if (detail.status !== 'ok') continue;
        const m = (detail.body as MessageMetadata | null) ?? {};
        if (!m.id) continue;
        messages.push({
          id: m.id,
          threadId: m.threadId ?? m.id,
          snippet: m.snippet ?? '',
        });
      }
      // listResult.status is 'ok' here (errored earlier).
      return listResult.status === 'ok' && listResult.truncated
        ? { status: 'ok', messages, truncated: true }
        : { status: 'ok', messages };
    },
  });
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/tools/list_inbox.py
"""`list_inbox` — sub-agent read tool. Returns id+threadId+snippet
summaries (no body). The sub-agent calls `get_message` per id when it
wants full content.
"""

from __future__ import annotations

import asyncio
from typing import Any

from lifecoach_agent.workspace_agent.run_gws import RunGwsOk, run_gws
from lifecoach_agent.workspace_agent.tools._deps import WorkspaceToolDeps

LIST_INBOX_TOOL_NAME = "list_inbox"


def _build_inbox_query(*, unread_only: bool, since: str) -> str:
    """Assemble the Gmail search query for triage.

    `in:inbox` intentionally excludes archived / moved mail (the old
    `label:INBOX` form leaked messages that had been archived). The
    `newer_than` term preserves the requested recency window, and
    `is:unread` narrows further when the caller asks for it.
    """
    terms = ["in:inbox"]
    if unread_only:
        terms.append("is:unread")
    terms.append(f"newer_than:{since}")
    return " ".join(terms)


def _dedupe_message_ids(messages: Any) -> list[str]:
    """Return the distinct message ids in first-seen order.

    Gmail's `users.messages.list` can surface the same id more than once
    (paging / query overlap); de-duping here means triage reads each
    message at most once instead of issuing a redundant `get` per copy.
    """
    seen: set[str] = set()
    ids: list[str] = []
    for message in messages or []:
        if not isinstance(message, dict):
            continue
        mid = message.get("id")
        if not isinstance(mid, str) or not mid or mid in seen:
            continue
        seen.add(mid)
        ids.append(mid)
    return ids


def create_list_inbox_tool(deps: WorkspaceToolDeps) -> Any:
    async def list_inbox(
        unread_only: bool = False, since: str = "1d", limit: int = 15
    ) -> dict[str, Any]:
        """List recent inbox messages as id+threadId+snippet summaries
        (no body). Use get_message per id to read full content. Read-only.

        Args:
            unread_only: When true, restrict to unread messages only.
            since: Gmail-style relative window (e.g. "1d", "12h", "3d").
                Default "1d" — last 24 hours.
            limit: Maximum number of messages to return (1–50). Default 15.
        """
        q = _build_inbox_query(unread_only=unread_only, since=since)
        max_results = max(1, min(int(limit), 50))

        list_result = await run_gws(
            store=deps.store,
            uid=deps.uid,
            tool_name=LIST_INBOX_TOOL_NAME,
            service="gmail",
            resource="users.messages",
            method="list",
            params={"userId": "me", "q": q, "maxResults": max_results},
            build_client=deps.build_client,
            log=deps.log,
        )
        if not isinstance(list_result, RunGwsOk):
            return {"status": "error", "code": list_result.code, "message": list_result.message}

        body = list_result.body if isinstance(list_result.body, dict) else {}
        ids = _dedupe_message_ids(body.get("messages"))
        if not ids:
            return {"status": "ok", "messages": []}

        details = await asyncio.gather(
            *[
                run_gws(
                    store=deps.store,
                    uid=deps.uid,
                    tool_name=LIST_INBOX_TOOL_NAME,
                    service="gmail",
                    resource="users.messages",
                    method="get",
                    params={"userId": "me", "id": mid, "format": "metadata"},
                    build_client=deps.build_client,
                    log=deps.log,
                )
                for mid in ids
            ]
        )
        messages: list[dict[str, Any]] = []
        seen_detail_ids: set[str] = set()
        for detail in details:
            if not isinstance(detail, RunGwsOk):
                continue
            m = detail.body if isinstance(detail.body, dict) else {}
            mid = m.get("id")
            if not isinstance(mid, str) or mid in seen_detail_ids:
                continue
            seen_detail_ids.add(mid)
            messages.append(
                {
                    "id": mid,
                    "threadId": m.get("threadId") or mid,
                    "snippet": m.get("snippet") or "",
                }
            )
        out: dict[str, Any] = {"status": "ok", "messages": messages}
        if list_result.truncated:
            out["truncated"] = True
        return out

    from google.adk.tools import FunctionTool

    list_inbox.__name__ = LIST_INBOX_TOOL_NAME
    return FunctionTool(list_inbox)
